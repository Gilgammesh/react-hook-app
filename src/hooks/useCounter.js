import { useState } from "react";

export const useCounter = (initialValue = 10) => {
  const [state, setState] = useState(initialValue);

  const increment = (factor = 1) => {
    setState(state + factor);
  };

  const decrement = (factor = 1) => {
    setState(state - factor);
  };

  const reset = () => {
    setState(initialValue);
  };

  return {
    counter: state,
    increment,
    decrement,
    reset,
  };
};
