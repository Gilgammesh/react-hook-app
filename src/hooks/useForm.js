import { useState } from "react";

export const useForm = (initialValues = {}) => {
  const [formValues, setForm] = useState(initialValues);

  const handleInputChange = ({ target: { name, value } }) => {
    setForm({
      ...formValues,
      [name]: value,
    });
  };

  return [formValues, handleInputChange];
};
