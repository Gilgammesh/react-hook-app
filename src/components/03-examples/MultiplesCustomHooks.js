import React from "react";
import { useCounter } from "../../hooks/useCounter";
import { useFetch } from "../../hooks/useFetch";
import "./styles.css";

const MultiplesCustomHooks = () => {
  const { counter, increment } = useCounter(1);

  const { data, loading, error } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );
  const { quote_id, quote, author } = !!data && data[0];

  return (
    <div>
      <h1>Breaking Bad Quote</h1>
      <hr />

      {loading ? (
        <div className="alert alert-info text-center">Loading ...</div>
      ) : (
        <div className="blockquote text-right">
          <h4>{quote_id}</h4>
          <p className="mb-0">{quote}</p>
          <footer className="blockquote-footer">{author}</footer>
        </div>
      )}

      <button className="btn btn-success" onClick={() => increment(1)}>
        Siguiente quote
      </button>
    </div>
  );
};

export default MultiplesCustomHooks;
