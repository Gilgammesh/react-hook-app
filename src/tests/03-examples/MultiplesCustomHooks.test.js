import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import MultiplesCustomHooks from "../../components/03-examples/MultiplesCustomHooks";
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";
jest.mock("../../hooks/useFetch");
jest.mock("../../hooks/useCounter");

describe("Pruebas unitarias en el componente MultiplesCustomHooks.js", () => {
  useCounter.mockReturnValue({
    counter: 1,
    increment: () => {},
  });
  /***********************************************************************************************************/
  test("se debe mostrar el componente <MultiplesCustomHooks /> correctamente", () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<MultiplesCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });
  /***********************************************************************************************************/
  test("se debe mostrar cargando... mientras se espera la información de la peticion url", () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<MultiplesCustomHooks />);
    const div = wrapper.find("div.alert");

    expect(div.exists()).toBe(true);
    expect(div.text().trim()).toBe("Loading ...");
  });
  /***********************************************************************************************************/
  test("se debe mostrar la información de la peticion url en el componente", () => {
    useFetch.mockReturnValue({
      data: [
        {
          quote_id: 1,
          quote: "I am the danger!",
          author: "Walter White",
          series: "Breaking Bad",
        },
      ],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultiplesCustomHooks />);
    const div = wrapper.find("div.blockquote");

    expect(div.exists()).toBe(true);

    const h4 = wrapper.find("div.blockquote h4");
    const p = wrapper.find("div.blockquote p");
    const footer = wrapper.find("div.blockquote footer");

    expect(h4.text().trim()).toBe("1");
    expect(p.text().trim()).toBe("I am the danger!");
    expect(footer.text().trim()).toBe("Walter White");
  });
});
