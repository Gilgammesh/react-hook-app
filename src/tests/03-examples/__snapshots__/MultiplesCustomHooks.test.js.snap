// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas unitarias en el componente MultiplesCustomHooks.js se debe mostrar el componente <MultiplesCustomHooks /> correctamente 1`] = `
<div>
  <h1>
    Breaking Bad Quote
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading ...
  </div>
  <button
    className="btn btn-success"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;

exports[`Pruebas unitarias en el hook personalizado MultiplesCustomHooks.js se debe mostrar el componente <MultiplesCustomHooks /> correctamente 1`] = `
<div>
  <h1>
    Breaking Bad Quote
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading ...
  </div>
  <button
    className="btn btn-success"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
